content/dev/blog/analyzing-so-data.Rmd

```diff
@@ -0,0 +1,308 @@
+
+* if there is only one mode / cluster don't adjust the salaries. This is the simplest
+case, if there's no discrepancy there's no reason to correct.
+* if any mode /cluster of salaries only represented 10% of the population, don't adjust the
+salaries. We found that sometimes `mclust` would find very small clusters and
+that would produce odd results, we chose 10% here out of convenience, but a more
+principle approach would pick a better cut off.
+* if there was more than one mode / cluster of salaries, and the lowest group's salary
+multiplied by 12 was greater than the second lowest group's salary multiplied by two, then do not
+attempt to adjust the salaries. Sometimes `mclust` would find clusters, but they
+would actually be fairly tightly together. This was likely the result of other
+covariates like experience, but critically, we didn't want to multiply junior
+developers' salaries by 12 just because they form a distinct cluster separate
+from senior developers.
+* otherwise, multiply the lowest salary group by 12 to correct for apparent misreporting.
+
+Finally, we only applied this adjustment to countries where we had at least 90
+responses. This allowed us to be reasonably confident that there were enough
+salaries in the sample to detect distinct clusters.
+
+```{r, cache = TRUE}
+library(crplyr)
+library(mclust)
+adjFrame <- ds %>%
+    select(Salary, Country, Respondent) %>%
+    collect()
+
+adjustSalary <- function(df) {
+    res <- Mclust(df$Salary)
+    df$adjSalary <- df$Salary
+    df$Group <- res$classification
+    summaryDf <- df %>%
+        group_by(Group) %>%
+        summarise(mean = mean(Salary), n=n())
+    if(any(summaryDf$n < nrow(df) / 10)){
+        # if any cluster is smaller than 1/10 of the sample, don't adjust salaries
+        df$adjSalary <- df$Salary
+        return(df)
+    }
+    if(nrow(summaryDf) > 1 && mean(df$adjSalary[df$Group==1])*12 > mean(df$adjSalary[df$Group==2])*2){
+        # if the lowest cluster*12 is greater than the second lowest*2, then the
+        # correction is over-aggressive, so don't adjust salaries
+        df$adjSalary <- df$Salary
+        return(df)
+    } else if (nrow(summaryDf) == 1){
+        # if there is a single cluster, don't adjust the salaries
+        df$adjSalary <- df$Salary
+        return(df)
+    }
+    # otherwise, multiple group 1's salaries by 12
+    df <- df %>%
+        mutate(adjSalary = case_when(Group == 1 ~ Salary * 12,
+                                         TRUE ~ Salary))
+    return(df)
+}
+
+adjusted <- adjFrame %>%
+    filter(!is.na(Salary)) %>%
+    group_by(Country)  %>%
+    filter(n() >= 90) %>%
+    do(adjustSalary(.)) %>%
+    ungroup() %>%
+    right_join(adjFrame[, "Respondent"], by = "Respondent")
+```
+
+We can then upload the data back up to the Crunch app:
+
+```{r, cache = TRUE}
+
+ds$adjSalary <- adjusted$adjSalary
+knitr::include_graphics("images/adjSalary_screen.png")
+
```

Comment:
The point of this section isn't that we reproduced the income adjustment that someone else blogged about. The point is that you can do this with the data in Crunch, just as you would be able to if you had the dataset locally. And, the data is in the cloud, sharable to others that don't do R, and you have access to all of your data anywhere, etc.

And then the post will need a conclusion summing up how great and powerful the R integration with Crunch is, etc.